src/TensorNetwork.jl

```diff
@@ -59,7 +60,29 @@ struct TensorNetwork <: AbstractTensorNetwork
             length(unique(dims)) == 1 || throw(DimensionMismatch("Index $(ind) has inconsistent dimension: $(dims)"))
         end
 
-        return new(indexmap, tensormap, CachedField{Vector{Tensor}}())
+        return new(indexmap, tensormap, CachedField{Vector{Tensor}}(), Ref(true))
+    end
+
+    function TensorNetwork(tensors, check_index_sizes::Bool)
+        tensormap = IdDict{Tensor,Vector{Symbol}}(tensor => inds(tensor) for tensor in tensors)
+
+        indexmap = reduce(tensors; init=Dict{Symbol,Vector{Tensor}}()) do dict, tensor
+            for index in inds(tensor)
+                # TODO use lambda? `Tensor[]` might be reused
+                push!(get!(dict, index, Tensor[]), tensor)
+            end
+            dict
+        end
+
+        if check_index_sizes
+            for ind in keys(indexmap)
+                dims = map(tensor -> size(tensor, ind), indexmap[ind])
+                length(unique(dims)) == 1 ||
+                    throw(DimensionMismatch("Index $(ind) has inconsistent dimension: $(dims)"))
+            end
+        end
+
+        return new(indexmap, tensormap, CachedField{Vector{Tensor}}(), Ref(check_index_sizes))
```

Comment:
i prefer another a function like `set_unsafe_context!(tn::TensorNetwork, ctx)` that sets that field, rather than another constructor